sd_hf_serve/README.md

```diff
@@ -0,0 +1,191 @@
+
+Inference services need to be (1) reliable, (2) performant, and (3) cost-effective. Reliability involves quick and automatic recovering from hardware and software failures. Performant means low latency response to user requests that includes scaling compute accelerators seamlessly, quickly, and keeping resource utilization optimal and finally, releasing resources not needed.  
+
+This solution describes a `stabilityai/stable-diffusion-2-1-base` [inference pipeline](oci-image-build) that implements reliable, performant and cost-effective AWS services using EC2 Inferentia instances. CodePipeline orchestrates CodeBuild DLC-based container images that are sourced in github repository and pushed to ECR. We compile the model with a [K8s batch/v1 Job](sd21-512-compile-job.yaml) that stores it in S3. Next, we deploy [K8s apps/v1 Deployment](sd21-512-serve-deploy.yaml) that pulls the model from S3 and instanciate it from images stored in ECR. Finally, we deploy [networking.k8s.io/v1 Ingress](sd21-512-serve-ingress.yaml) that builds AWS LoadBalancer and TargetGroups that auto discover the pods that power the Inference app (Gardio). 
+![alt text](./sdhfserve.png)
+
+For reliability, we use Karpenter. Karpenter node-pool manages Inferentia node lifecycle. Karpenter adds Inferentia nodes to handle unschedulable pods, schedules pods on those nodes, and removes them when not needed. 
+
+Our [NodePool](inf2-nodepool.yaml) uses the instance-familiy key to specify the instances needed i.e., `inf2.xlarge`, `inf2.8xlarge`, `inf2.24xlarge`, and `inf2.48xlarge`. 
+
+```yaml
+       - key: karpenter.k8s.aws/instance-family
+          operator: In
+          values: ["inf2"]
+```
+
+The Pod resource limits will determine which Inf2 instance will be requested by the total number of required neuron devices. In the example below the scheduler will launch inf2.xlarge when the replica-set size is 1 and increase linearly by the total number of needed neuron devices. 
+
+```yaml
+        resources:
+          limits:
+            aws.amazon.com/neuron: 1
+```
+
+The neuron-device-plugin returns the list of Neuron cores/devices to kublet. Kubelet advertises the Core/Device list to K8s API server (in turn to Kube-Scheduler). Kube-Scheduler picks up the pod creation request. Kube-scheduler calls the neuron-scheduler-extn filter function with a list of nodes and pod specifications. neuron-scheduler scans the nodes and filters out nodes with non contiguous cores/devices and returns the nodes that are capable of supporting the pod specification. Kube-scheduler invokes the neuron-scheduler bind function with pod and node. It updates the pod annotation with allocated neuron contiguous core/device Ids. Then neuron-scheduler sends the bind request to kubelet of the selected node. Kubelet calls the Alloc function of the neuron-device-plugin. neuron-device-plugin queries the pod annotation for allocated core/device Ids and exports the device cores to container runtime. 
```

Comment:
in the neuron release 2.18 the scheduler will be slightly different for trn1. On inf2/inf1 it will filter out nodes with non-contiguous cores/devices but on trn1 there will be contiguous sets of devices/cores that are not allowed. The documentation will be updated for the 2.18 release https://github.com/aws-neuron/private-aws-neuron-sdk-staging/pull/1533/files

But you may want to add a comment linking to the documentation https://awsdocs-neuron.readthedocs-hosted.com/en/latest/containers/kubernetes-getting-started.html. Something like "to learn more about how the Neuron scheduler filters nodes see the documentation"

Reply:
Added reference to doc